Mage/src/main/java/mage/game/permanent/PermanentImpl.java

```diff
@@ -1913,6 +1915,32 @@ public boolean isRingBearer() {
         return ringBearerFlag;
     }
 
+    @Override
+    public boolean isSolved() {
+        return solved;
+    }
+
+    @Override
+    public boolean solve(Game game, Ability source) {
+        if (this.solved) {
+            return false;
+        }
+        GameEvent event = new GameEvent(GameEvent.EventType.SOLVE_CASE, getId(),
+                source, source.getControllerId());
+        if (game.replaceEvent(event)) {
+            return false;
+        }
+        Player controller = game.getPlayer(source.getControllerId());
+        if (controller != null) {
+            game.informPlayers(controller.getLogName() + " solves " + this.getLogName());
```

Comment:
There are useful `main log text + CardUtil.getSourceLogName(game, source)` to show source of the event and workable popup hint. It's a standard way for all log events.